### PR TITLE
Dynarrays, unboxed (with local dummies)

### PR DESCRIPTION
**Pull request by gasche**
We recently merged Dynarray in the stdlib (yay! #11882 ), with the caveat that its implementation is 'boxed', it uses a representation similar to `'a option array` to safely represent 'empty' values without leaking user data.

#11882 started its life as an attempt to un-block @c-cube's #11563, the previous proposal for Dynarray in the stdlib, which used an 'unboxed' representation. The PR discussion had ground to a halt because we disagreed on how which unsafe tricks to use to implement this unboxed approach.

The present PR proposes to move Dynarray to an unboxed representation, using one of the two approaches discussed in #11563.so there will be at least one release with boxed dynarrays.

The design was in particular informed by discussions with @lthls and @xavierleroy.
I was motivated to work on this now by @backtracking's interest in building efficient data structures on top of Dynarray (see #12871).

Timing remark: Dynarray will be released with OCaml 5.2, but the present PR might be merged in 5.3 at best, so there will be at least one release with boxed dynarrays. Boxed dynarrays are just fine.

#### How to review

I think that correctness is the main question when reviewing this PR. It should be clear that the implementation is indeed unboxed, and I think that the performance benefits of this choice are well-understood. On the other hand, it is not clear that the implementation is safe/sound -- that the user can never observe a type-incorrect dummy -- and this is what needs reviewing foremost.

## Performance

The performance of the unboxed version are *generally* better, especially on very large arrays (one million elements). For example, on a microbenchmark that starts from an empty dynarray and repeatedly adds 1000 elements and pops them again, we get the following results:

```
  Stack ran
    1.11 ± 0.05 times faster than Queue
    1.22 ± 0.05 times faster than CCVector
    1.33 ± 0.05 times faster than Dynarray_unboxed
    1.59 ± 0.06 times faster than Dynarray_boxed
    1.61 ± 0.06 times faster than Base_stack
```

where `Stack` and `Queue` are the stdlib modules, `CCVector` is the dynarray of Containers, `Dynarray_boxed` is the current stdlib implementation, `Dynarray_unboxed` is our new implementation, and `Base_stack` is the `Stack` module of Base. (Both Base and Containers use an unboxed representation.)

Please keep in mind that those are *small* performance differences for a benchmark that measures only the data structure and no useful user work: for containers of size 1000, Array is 2x faster than all those Dynarray implementations at random access, and Hashtbl is 42x slower than Array.

The main advantage of unboxed representations is to avoid surprises regarding memory layout and usage, and to benefit from better locality. Locality effects are hard to measure, especially in microbenchmarks, so I did not try to do so -- let us keep in mind that there are probably more performance benefits in real-world programs than shown above. I did a more representative study of the impact of boxing on a real-world program in the previous PR ( https://github.com/ocaml/ocaml/pull/11882#issuecomment-1405867624 ), where making one of @c-cube's implementations boxed caused a 25% slowdown at worst on some dynarray-critical programs.

## Dummies

A dynarray that contains `length` user values is represented by a "backing array" of length `capacity`, with `capacity >= length`. The positions from `0` to `length - 1` are the "user space", storing the values provided by the user, and the positions from `length` to `capacity - 1` are the "empty space". Adding a new element increments the mutable field `length`, increasing the user space and shrinking the empty space.

An unboxing representation is achieved by using a 'dummy' in the empty space. With OCaml 5, we cannot guarantee that the bound check on the current `length` whether a given element is a dummy or not: there could be data races on the dynarray that result in reading dummy values in the user space. So we need to find a 'dummy' that we can distinguish at runtime from any user value -- in particular the old trick of using `Obj.magic ()` does not work.

In #11563 we discussed two approaches for dummies:

1. Local dummies, where we allocate a private reference (we never show the value to anyone), and we use this reference as the dummy, using physical equality to check whether an element is the dummy or not.
2. Atomic dummies, where we use "atoms" of non-zero tag, which are low-level values expressible in the OCaml runtime representation but not currenty used to represent any OCaml value.

If you want the details, the discussion of these approaches starts at https://github.com/ocaml/ocaml/pull/11563#issuecomment-1257271948 and there is a lot of it. The summary is that both approaches have downsides, but (2) got a fairly strong veto from Xavier, so I think that (1) is a safer bet if we want an unboxed representation. The present PR implements 1.

I don't expect any performance implication to using (1) or (2), except for the following. The implementation of (1) makes dynarrays 3-field records instead of 2-field records, and allocates a dummy value on each new dynarray creation, so it may be slightly slower when creating a lot of extremely small dynarrays.

### Details on the current implementation

If you are an expert, you may wonder how the approach deals with marshalling, and how it avoids issues coming from flat float arrays. The answer to both these questions is in the long implementation comment at the beginning of stdlib/dynarray.ml. Go review my code!

### A type-rich API for dummies

Initially I implemented dummies using `Obj.magic` and that's it: you have an `'a array`, but some values are not in fact valid `'a` values and you better be careful about it -- forget a `v == Obj.obj dummy` check and that's a segfault for your users.

Then I realized that I could hide the definition of dummies inside a submodule enforcing a type discipline on the use of dummies. Something like:

```ocaml
type dummy
val fresh : unit -> dummy

type 'a or_dummy
val of_val : 'a -> 'a or_dummy
val of_dummy : dummy -> 'a or_dummy
val find : 'a or_dummy -> 'a option
```

In fact you can do even better, by using a `'stamp` type parameter on dummies, to give a different type to two distinct dummy values. (This is also called "branding" sometimes.)

```ocaml
type 'stamp dummy
type fresh_dummy = Fresh : 'stamp dummy -> fresh_dummy
val fresh : unit -> fresh_dummy

type ('a, 'stamp) or_dummy
val of_val : 'a -> ('a, 'stamp) or_dummy
val of_dummy : 'stamp dummy -> ('a, 'stamp) or_dummy
val find : ('a, 'stamp) or_dummy -> 'a option
```

This is exactly the sort of improductive type over-engineering that I like, so of course I implemented it. And it *caught a bug* in my code.

The bug was in `append : 'a Dynarray.t -> 'a Dynarray.t -> unit`, `append a b` adds all the elements of `b` to `a`. I implemented this by using `Array.blit` to blit the elements of `b`'s backing array, but this is unsound because `a` and `b` may use different dummy values, so you have to carefully fail on dummy elements in `b`'s user space. This became a hard type error with the above strongly typed API.

(Besides the over-engineered typing, which has upsides and downsides, I like the fact that this approach forced me to isolate all the unsafe code and low-level reasoning in a submodule with a clear boundary from the rest of the Dynarray code.)

### CODE REVIEW

File: stdlib/dynarray.ml
Comment (c-cube) [inline stdlib/dynarray.ml:197]: should this be `[@@unboxed]` too?

File: stdlib/dynarray.ml
Comment (c-cube) [inline stdlib/dynarray.ml]: There could still be a global dummy that is used by all dynarrays but the unmarshalled ones, I think? That would save some allocations on each dynarray creation.

File: stdlib/dynarray.ml
Comment (c-cube) [inline stdlib/dynarray.ml:215]: I've been wondering if `let[@inline]` has an impact on this kind of code even without flambda. If it does, I think many of these functions should be annotated because we really don't want any overhead in dummy<->real value conversion.

**Review by c-cube — COMMENTED** [review]
I like this! It's a bit more elaborate than what I had in containers, but it gains the good marshalling properties and it's more robust anyway, I think. Thank you @gasche :)

File: stdlib/dynarray.ml
Comment (gasche) [reply stdlib/dynarray.ml:197]: Good question :-) At the level of the signature, `'stamp dummy` is abstract, and unboxing `Fresh` is forbidden:

```
File "/home/gasche/Prog/ocaml/github-trunk/stdlib/dynarray.ml", line 114, characters 2-68:
114 |   type fresh_dummy = Fresh : 'stamp dummy -> fresh_dummy [@@unboxed]
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: This type cannot be unboxed because
       it might contain both float and non-float values,
       depending on the instantiation of the existential variable "'stamp".
```

I *think* that the error is over-conservative -- we should be able to relax the check, accept unboxing in the signatures as long as all implementations use a sound concrete choice of `dummy`. We could also make it go away by exposing the definition of `dummy`, fully or in part. But neither of this is necessary as the representation of `Fresh` is not performance-sensitive.

File: stdlib/dynarray.ml
Comment (gasche) [reply stdlib/dynarray.ml]: I had the same idea while writing the description of the PR, and I pushed a commit to implement this -- but I forgot to update this bit of the documentation, thanks!

File: stdlib/dynarray.ml
Comment (gasche) [reply stdlib/dynarray.ml:215]: I checked that the `Dummy` module does not introduce any inefficiency in the `ocamlopt`-generated code, even without flambda. (I checked by comparing the -dcmm output before and after the commit that introduces this typed API, they are the same except for the bugfix in `append`.)

File: stdlib/dynarray.ml
Comment (c-cube) [reply stdlib/dynarray.ml:197]: Now that dummies are allocated only once (ignoring Marshal) I think it's not an issue anyway, indeed! :)

File: stdlib/dynarray.ml
Comment (yannl35133) [inline stdlib/dynarray.ml]: Why leave the length check if `Array.make` is already called ?
Did you mean to call it after the check ?

Comment (gasche) [reply stdlib/dynarray.ml]: I think so, indeed. Thanks!

**Comment by alainfrisch** [comment]
It might not fit the current abstraction, but did you consider using as the dummy value the array itself? It might behave slightly better performance-wise (locality), while avoiding the extra slot in the record (and still behave ok wrt marshaling, with the same cycle behavior with No_sharing).

**Comment by gasche** [comment]
Xavier suggested it as well, but it is technically not correct for all OCaml setups: with `-rectypes` I can put the value of the array in itself.

**Comment by alainfrisch** [comment]
> Xavier suggested it as well, but it is technically not correct for all OCaml setups: with `-rectypes` I can put the value of the array in itself.

But the array is not directly exposed, only a record containing it (plus the current length), no?

**Comment by alainfrisch** [comment]
To be more explicit : I was thinking of using the backing array as the dummy value (which makes resizing a tiny bit more complex).

**Comment by gasche** [comment]
Using the backing array is not very convenient because it is mutable. Currently to grow the backing array I can allocate a new array and then blit the elements of the current array. If we used the array as dummy, I would need an extra traversal to check for the absence of "old" dummy values after the blit. Same thing for operations that currently use `Array.sub` (`copy` and `{fit,set}_capacity`). I think that the performance overhead of these extra traversals offsets the benefits of saving one word per dynarray.

**Comment by alainfrisch** [comment]
Indeed, that's what I meant by "making the resizing more complex". Basically, one needs a blit operation that preserves "pointers to self"; this could be done in one linear pass, but perhaps it's too costly indeed.

**Comment by alainfrisch** [comment]
Another idea to avoid a dummy field per array  : what about using a global dummy, which is preserved across marshaling? This could be achieved with a custom block (and marshaling/demarshaling operations). This could also avoid problems with the generic comparison (even if we don't want to encourage using them, going into an infinite loop is not so good).

**Comment by gasche** [comment]
You have a good point that polymorphic comparison is broken by the use of a cyclic dummy, when comparing two structurally-equal dynarrays with distinct dummies. (If they are distinct comparison will stop on the array before looping; if the dummies are equal then the `if (a == b) return true` fast-path will work.)

**Comment by alainfrisch** [comment]
Also : The fast path on physical equality is only for Stdlib.compare, not for (=).

**Comment by gasche** [comment]
I published more detailed microbenchmark results in [dynarray-benchmarks/BENCH.md](https://gitlab.com/gasche/dynarray-benchmarks/-/blob/trunk/BENCH.md?ref_type=heads), in case someone wants to look at the gory details. My summary of the results would be as follows:

- For stack-like usage, the list-based `Stack` is noticeably faster for small to medium-sized data, but noticeably slower for very large data (a million elements); Stack is better than Dynarray for most situations that do not need random access.
- The performance of the boxed version of dynarray is broadly the same as the unboxed version for element-wise operations; but it is noticeably slower for array-based operations (`to_array`, `of_array`) where the use of an unboxed array allows to use runtime-optimized Array functions (`copy`, `sub`).
- The unboxed Dynarray implementation has broadly the same performance as the other unboxed-dynarray implementations we compared to (CCVector and Base's Stack). 
   In particular, the design choices that are specific to Dynarray (reducing unsafe operations to be safe with OCaml 5, checking against iterator invalidation) do not seem to negatively affect performance on the operations that we tested.

**Comment by gasche** [comment]
@alainfrisch I am trying to avoid bad interactions between the cyclic dummy (for marshalling) and comparison by... hiding the cycle inside an object. It feels like fighting fire with fire, but it appears to work, I pushed a commit implementing this approach.

Writing a testsuite for this forced me to realize that comparison for dynarray values is neither fully structural neither by-identity: some physically-distinct but structurally-equal dynarrays will be considered equal, some will be considered distinct. (The most common reason for distinctness is: backing arrays have a different size. The less common is the use of different dummies.)
I think that this is okay, but it should be documented explicitly.
 
I could also ensure that equality is purely by-identity by sticking a unique identifier in each dynarray. This gives clearer specifications, but it also has a small cost on dynarray creation (only observable, if at all, on very small dynarrays). I am not sure whether this would be better than the current status.

**Comment by alainfrisch** [comment]
@gasche : this starts feeling both heavy (possibly impacting performance if we use many tiny dynarrays) and overly complex.

What about my proposal of using singleton custom blocks to create a global "dummy" value, which is properly restored upon demarshalling?  This would even work with the No_sharing flag, btw (not that we'd recommend doing that). We could even decide to make generic comparison fail on that dummy value, which is clearly better than looping forever, and perhaps even better than the current solution (because generic comparison on dynarray is not really meaningful anyway).

This solution would require exposing a runtime primitive to create that dummy value, and declare/call that value from Dynarray's implementation (at init time). As an extra safety net, one could make sure the primitive fails if called a second time, to make sure this value really cannot be obtained by the user (and stored in a Dynarray) -- but I'm not sure this check is really needed.

What are the downsides of this approach, apart from requiring a tiny bit of runtime support (and just an adjustment to js_of_ocaml as well)?

A more general approach (and thus possibly easier to justify changing the runtime) would be to expose:
````
val Obj.mk_dummy: string -> Obj.t
````
(again, possibly with a runtime failure if the function is called several times on the same string)

This would allow implementing other data structures with similar needs, without the risk of interfering with `Dynarray`. Each such data structure would just need to use a globally unique name for its dummy value and make sure to never expose that value (and a globally unique string is easier than a globally unique atom tag number).

I'm thinking for instance of a data structure that would expose a raw "array of unboxed options" (btw, perhaps we should do that first, and use it to implement the backing array of Dynarray -- except that the "clean" API for "array of unboxed option" might not play well with Dynarray, performance wise).  Or the simpler "reference on unboxed option" (which is really a special case of "array of unboxed options" of size 1).

**Comment by gasche** [comment]
The dummy is shared between all dynarrays created by the program, so the empty-object creation happens only once at runtime. (Unmarshalling may create new dummies on the fly; so there is a cost when unmarhsalling small dynarrays, but I'm fine with that.)

Your proposal of using a custom singleton is interesting, but much more invasive as a change to my current implementation, so I went for my simpler approach for now.

I don't want to help other people create their own dummies in the context of this PR, or generalize Dynarray to another data structure of unboxed array with holes, because I think that this is likely to create new design problems and derail progress on making Dynarray unboxed.

**Comment by alainfrisch** [comment]
> Your proposal of using a custom singleton is interesting, but much more invasive as a change to my current implementation, so I went for my simpler approach for now.

It's invasive in the sense that it requires a bit of runtime support (but it's straightforward), but it would simplify the OCaml side quite a bit, by avoiding the need to thread the `dummy` through many functions.

> generalize Dynarray to another data structure of unboxed array with holes

I was not proposing to generalize Dynarray, but to introduce another data structure (array of unboxed options), which could serve to *implement* the Dynarray backing array on top of it (without changing the API of Dynarray).  The other data structure would have its own direct uses.  And implementing that other data structure would really benefit from a global singleton (otherwise one needs to turn the internal array into a record, adding an extra indirection on each access, which is really not nice).

**Comment by alainfrisch** [comment]
(Of course, one could also decide that it's ok to go with your proposal, and possibly switch to a global dummy later, but then we'd break the Marshal-level compatibility. My opinion is that it's worth exploring other options now, but I'm in the easy position of the commentator not doing the actual work, so just take my comment as comment, not as a strong opposition to this PR.)

**Comment by gasche** [comment]
@damiendoligez, @OlivierNicole : you both worked on #12889, I wonder if you may be interested in looking at this one as well -- it is more work.

(I don't disagree with @alainfrisch that it would be worth exploring a runtime-supported dummy generator, but I would like to amortize my work by waiting for this PR to make progress before that. I don't think it would be a real issue if this runtime-dummy mechanism had to wait until a later release.)

**Comment by OlivierNicole** [comment]
Sure, I can look at this.

File: stdlib/dynarray.ml
Comment (OlivierNicole) [inline stdlib/dynarray.ml]: I’m not sure to see why the locally abstract type is useful here.

**Comment by OlivierNicole** [comment]
I am about halfway in reviewing this; but also, to test the claim that the code should never crash, even in presence of erroneous parallel use, I exercised (part of) the API using [multicoretests](https://github.com/ocaml-multicore/multicoretests/), stress-testing it with parallel calls to `get_last`, `add_last`, but also `set_capacity`, `reset`… and at no point over dozens of thousands of tests did it crash! This is nice.

File: stdlib/dynarray.ml
Comment (OlivierNicole) [inline stdlib/dynarray.ml]: `let {arr; length; _} = a` would avoid spurious “unmatched field” warnings (warning which I find useful when modifying types, especially in difficult code).

File: stdlib/dynarray.ml
Comment (OlivierNicole) [inline stdlib/dynarray.ml:639]: I know this is previously reviewed code, but why not `List.iter (add_last a) li`?

File: stdlib/dynarray.ml
Comment (OlivierNicole) [inline stdlib/dynarray.ml:861]:
```suggestion
      p (unsafe_get arr ~dummy ~i ~length)
      || (loop [@tailcall]) p arr dummy (i + 1) length
```
I assume the lazy or operator preserves tail recursion, but doesn’t it soothe the mind to enforce it?

File: stdlib/dynarray.ml
Comment (OlivierNicole) [inline stdlib/dynarray.ml:874]:
```suggestion
      p (unsafe_get arr ~dummy ~i ~length)
      && (loop [@tailcall]) p arr dummy (i + 1) length
```

File: stdlib/dynarray.ml
Comment (OlivierNicole) [inline stdlib/dynarray.ml]: This comments should be updated, it refers to the boxed representation.

File: stdlib/dynarray.ml
Comment (OlivierNicole) [inline stdlib/dynarray.ml:72]: A related but different danger is that care should be taken to never store naked floats inside a backing array, as the GC would attempt to scan them.

File: stdlib/dynarray.ml
Comment (OlivierNicole) [inline stdlib/dynarray.ml]: This does not work well if `a` is a float array: `Array.blit` will see it has a `Double_array_tag` and assume the destination array is a float array as well (which the typechecker enforces in the absence of unsafe casts), and copy the naked floats into the backing array without boxing them. We now have a 0-tagged array with naked floats, which will crash the GC. And I did get segfaults with things like `Dynarray.of_list [ 1.; 2.; 3. ]`. `Dynarray.of_array` is affected too.

I wish I had found this myself when reading the code, but no, I stumbled on it in the process of writing a test for [qcheck-stm](https://github.com/ocaml-multicore/multicoretests).

**Review by OlivierNicole — COMMENTED** [review]
I like a lot the way this is written, the use of types for improving safety around tricky operations is very pleasant.

That being said, it is still very tricky code with subtle invariants to maintain in subtle ways. Those invariants being the validity of indices into the backing array (with a lot of `Array.unsafe_{get,set}`) and the absence of naked floats in said array; all of which must be maintained even in the presence of data races. I found one bug during testing (see specific comments), and although I have been extra careful and did not see other problems, I don’t fully trust myself.

So for extra confidence, I am in the process of writing a full test for this module using multicoretests, including a parallel stress test verifying the absence of crashes. Once that runs smoothly (and perhaps after a bit more eyeballing of the operations around floats), I am fine with approving.

File: stdlib/dynarray.ml
Comment (OlivierNicole) [reply stdlib/dynarray.ml]: Never mind, this code has been rewritten.

File: stdlib/dynarray.ml
Comment (OlivierNicole) [reply stdlib/dynarray.ml]: Never mind, this code has been rewritten.

File: stdlib/dynarray.ml
Comment (gasche) [reply stdlib/dynarray.ml]: Ah, so it looks like you found a bug due to incorrect assumptions about `Array.blit`.

I'm not sure that the explanation of the bug is correct: the implementation of `blit` in [runtime/array.c:caml_array_blit](https://github.com/ocaml/ocaml/blob/d0ac496c2dc3e174f2bfff8ad78da1f398c24921/runtime/array.c#L382) does not make the assumption that you describe -- or at least I am not seeing it in the code. In the code I see a check on whether the destination array is a float array, but no check on the source array. I think that was is going on is that the non-flat-float-array implementation is incorrect if the source array is a flat float array.

In any case, your conclusion is correct that I should not use `Array.blit` here, I will update the code. Thanks!

Comment (OlivierNicole) [reply stdlib/dynarray.ml]: >  I think that was is going on is that the non-flat-float-array implementation is incorrect if the source array is a flat float array.

Ah yes, you’re right. I would complete: if the source array is a flat float array and the destination array is not, which of course is not supposed to happen.

Comment (gasche) [reply stdlib/dynarray.ml]: I pushed a new commit that should fix this issue by avoiding all uses of `Array.blit` on standard arrays unless they are known to not be flat float arrays. (There remains a direct call in `Dummy.Array.extend`, but the type tells us that the source array is one of our own non-flat arrays.)

Of course, this would be vastly easier if we could rely on first-class support for "uniform arrays", as proposed in https://github.com/ocaml/RFCs/pull/37 . (Or maybe, the complexity would be moved from Dynarray, where it feels accidental, to the Uniform_array module, where it makes sense to have it.) I don't have the energy to push for this right now, but hopefully we will have them at some point and be able to simplify the code here.

**Comment by gasche** [comment]
I need to go over the other comments in @OlivierNicole's review, and rebase the PR on top of the current trunk. Thanks!

File: stdlib/dynarray.ml
Comment (gasche) [reply stdlib/dynarray.ml:72]: It is really really hard to write a naked float into an array, and I think that this could in fact be considered a bug in the runtime that we managed to do it using `Array.blit`. I don't think it is worth mentioning here -- but I will go add an assertion in `caml_array_blit` against this.

File: stdlib/dynarray.ml
Comment (OlivierNicole) [reply stdlib/dynarray.ml]: Yes, the trickiness of the Dynarray code makes me even more convinced of the relevance of uniform arrays.

File: stdlib/dynarray.ml
Comment (gasche) [reply stdlib/dynarray.ml:639]: I don't remember, but I think that this made a noticeable difference in benchmarks. (If that was the reason, it would have been nice to write it down in a comment.)

**Comment by OlivierNicole** [comment]
Wouldn’t it be better to have `Dynarray.equal` and `Dynarray.compare`? It is appreciable to have non-polymorphic comparison functions for standard types, and in addition neither `(=)` nor `(==)` express equality of the elements.

**Comment by gasche** [comment]
This is a good idea, but it does not belong to the current PR. You could open an issue (or a PR!) for this. I'm happy to implement it when I get the time.

**Comment by OlivierNicole** [comment]
Ah, yes, done at #13104.

**Comment by OlivierNicole** [comment]
In addition to my review, I was not able to make the last version of the Dynarray module crash using parallel, randomized property testing, despite trying rather hard. So I’m fine with approving after a rebase.

**Comment by gasche** [comment]
I went over @OlivierNicole's comments and rebased the PR against trunk. I will merge if the CI agrees.

**Review by OlivierNicole — APPROVED** [review]
Looks good to me, as discussed.

**Comment by gasche** [comment]
Thanks again to all reviewers and in particular @OlivierNicole, this was a tricky one.

Things that would be worth doing as follow-up work:
- implement `compare` and `equal`
- introduce support in the runtime for "fresh singletons" to be used as dummy values (a data-less custom block with a unique identity, whose uniqueness is preserved by serialization), without using a recursive value wrapped inside an object
- introduce a minimal amount of support for uniform arrays in the runtime and stdlib, and rework Dynarray to use that instead of rolling our own

**Pull request merged**
